### PR TITLE
*: fix union scan for partitioned table

### PR DESCRIPTION
**Pull request by tiancaiamao**
<!--
Thank you for contributing to TiDB! Please read TiDB's [CONTRIBUTING](https://github.com/pingcap/tidb/blob/master/CONTRIBUTING.md) document **BEFORE** filing this PR.
-->

### What problem does this PR solve? <!--add issue link with summary if exists-->

```
MySQL [test]> show create table t1;
+-------+---------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------+
| Table | Create Table                                                                                                                                                                                                                                                                                                                          |
+-------+---------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------+
| t1    | CREATE TABLE `t1` (
  `c1` smallint(6) NOT NULL,
  `c2` char(5) DEFAULT NULL
) ENGINE=InnoDB DEFAULT CHARSET=utf8mb4 COLLATE=utf8mb4_bin
PARTITION BY RANGE ( `c1` ) (
  PARTITION p0 VALUES LESS THAN (10),
  PARTITION p1 VALUES LESS THAN (20),
  PARTITION p2 VALUES LESS THAN (30),
  PARTITION p3 VALUES LESS THAN (MAXVALUE)
) |
+-------+---------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------+
1 row in set (0.00 sec)

MySQL [test]> begin;
Query OK, 0 rows affected (0.00 sec)

MySQL [test]> insert into t1 values (1,1);
Query OK, 1 row affected (0.00 sec)

MySQL [test]> select * from t1;
Empty set (0.01 sec)
```

### What is changed and how it works?

How does union scan works for partitioned table:

1. In `rule_partition_processor.go`, the rewrite rule changes

```
UnionScan->DataSource
```

to

```
UnionAll{UnionScan->Partition1, UnionScan->Partition2 ...}
```

Note that UnionScan is now the parent of every Partition, rather than the previous

```
UnionScan->UnionAll{Partition1, Partition2 ...}
```

2. UnionScan->Partition1 transformed to UnionScanExec+three reader, the reader read data from partition, and UnionExec **get the incremental buffer changes on partition, not table**

3. To get the incremental buffer changes on partition:

* The id should be physicalID, instead of tableID
* Dirty table is organised using partition, i.e, `DirtyDB = map(physicalID -> changes)`

4. In the three kinds of reader(TableReader,IndexReader,IndexLookup) struct, `table` is also changed to partition (`PhysicalTable`)

5. The function signature `StmtAddDirtyTableOP` should be 

```
StmtAddDirtyTableOP(op int, physicalID int64, handle int64, row []types.Datum)
```

instead of 

```
StmtAddDirtyTableOP(op int, tid int64, handle int64, row []types.Datum)
```

<!-- Reviewable:start -->
---
This change is [<img src="https://reviewable.io/review_button.svg" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/pingcap/tidb/8871)
<!-- Reviewable:end -->

### CODE REVIEW

**Comment by tiancaiamao** [comment]
PTAL @lamxTyler @zz-jason

**Comment by tiancaiamao** [comment]
PTAL @zz-jason @lamxTyler

File: executor/builder.go [hunk 1]
```diff
@@ -771,6 +771,16 @@ func (b *executorBuilder) buildUnionScanFromReader(reader Executor, v *plannerco
 		us.conditions = v.Conditions
 		us.columns = x.columns
 		err = us.buildAndSortAddedRows()
+	case *UnionExec:
+		// For partitioned table, table reader may be transformed to UnionAll.
+		if x.fromPartition != nil {
+			us.dirty = GetDirtyDB(b.ctx).GetDirtyTable(x.fromPartition.Meta().ID)
+			us.conditions = v.Conditions
+			us.columns = x.fromPartition.Meta().Cols()
```
[the remark this answers is not in the document]
Comment (zz-jason) [reply executor/builder.go]: @tiancaiamao Please take a look at this comment.

File: planner/core/physical_plans.go [hunk 2]
```diff
@@ -269,6 +270,7 @@ type PhysicalLimit struct {
 // PhysicalUnionAll is the physical operator of UnionAll.
 type PhysicalUnionAll struct {
 	physicalSchemaProducer
+	FromPartition table.Table
```
Comment (winoros) [inline planner/core/physical_plans.go]: Then this field is meaningless

File: executor/builder.go (hunk 1 above)
Comment (winoros) [inline executor/builder.go]: What about
- Check whether the child operator of `UnionScan` is `UnionAll`.
- If so, use its `Children[0]` instead.
- If not, just use itself.
- Use the `switch` clause of line 703.

**Comment by zz-jason** [comment]
@tiancaiamao any update?

**Comment by tiancaiamao** [comment]
I'll do a tiny refactor to remove the physicalID from three reader after this PR.

File: executor/union_scan_test.go [hunk 3]
```diff
@@ -77,6 +77,20 @@ func (s *testSuite2) TestDirtyTransaction(c *C) {
...
+	tk.MustExec(`CREATE TABLE t (c1 smallint(6) NOT NULL, c2 char(5) DEFAULT NULL) PARTITION BY RANGE ( c1 ) (
+			PARTITION p0 VALUES LESS THAN (10),
+			PARTITION p1 VALUES LESS THAN (20),
+			PARTITION p2 VALUES LESS THAN (30),
+			PARTITION p3 VALUES LESS THAN (MAXVALUE)
+	)`)
+	tk.MustExec("begin")
+	tk.MustExec("insert into t values (1, 1)")
+	tk.MustQuery("select * from t").Check(testkit.Rows("1 1"))
+	tk.MustQuery("select * from t where c1 < 5").Check(testkit.Rows("1 1"))
```
Comment (lamxTyler) [inline executor/union_scan_test.go]: Could you add the test mentioned in https://github.com/pingcap/tidb/pull/8871#discussion_r244658362?

Comment (tiancaiamao) [reply executor/union_scan_test.go]: Done

**Comment by tiancaiamao** [comment]
PTAL @lamxTyler

**Comment by crazycs520** [comment]
@tiancaiamao  Need cherry pick to v2.1.

**Pull request merged**